### DISCUSSION TITLE
tsh login no browser dependency

### DISCUSSION THREAD

**Discussion by wmendes-ionos** [gravitational/teleport#16784] [Q&A] [answered]
I am aware of --browser=none functionality. Also that it is possible to use that on a remote host via tsh ssh port forwarding. However, is that possible to programatically authenticate ? Literally without opening a browser.

I've tried calling the tsh login link with curl -L but it didn't do the trick.

Also, is there a way to tsh login output only the link, without the message ?

**Comment by webvictim** [comment, marked as answer]
This isn't possible currently due to the large variety of different SSO providers out there and the different ways they handle authentication. I believe there is work going on to allow terminal logins specifically from certain providers without a browser, but at the moment the only way is to use SSH port forwarding, or log in on a local machine and use SSH agent forwarding to connect to a remote machine and make your Teleport certificates available.

You should also be able to use `tsh ssh -J bastion-hostname node-hostname` to use an intermediate host as a jump server to connect to a specific node if you need.

**Comment by bsgrigorov** [comment]
is there any update or workaround (that is not too hacky) on this? Specifically I would like to access my kube clusters from a shell on my phone or from github codespaces or from a no UI VM.

**Reply by webvictim** [reply]
We're working on `tsh kube login --headless` for these sorts of use cases. It should be released in 15.0 early next year.

**Reply by bsgrigorov** [reply]
Is there a workaround I could do currently with tbot or something else?

**Reply by webvictim** [reply]
`tbot` doesn't work with SSO users (and presumably using an SSO user is the reason why you need a browser at all)

You could use `tbot` with a local user, however - or export a longer lived identity file for that user with `tctl auth sign`.

**Reply by bsgrigorov** [reply]
if I use tbot do I need to be able to use the Yubikey? What does the setup look like such that credentials are not long lived? Yubikey usage also is not possible in a github codespace or VM.

**Reply by webvictim** [reply]
No, if you join a bot it does not require the use of a second factor.
